### PR TITLE
 PUBDEV-5837: Expose terminal node ids in R & Python

### PR DESCRIPTION
**Pull request by michalkurka**

### CODE REVIEW

File: h2o-core/src/main/java/water/api/ModelMetricsHandler.java [hunk 1]
```diff
@@ -468,7 +471,8 @@ public ModelMetricsListSchemaV3 predict(int version, ModelMetricsListSchemaV3 s)
         assert(Model.LeafNodeAssignment.class.isAssignableFrom(parms._model.getClass()));
         if (null == parms._predictions_name)
           parms._predictions_name = "leaf_node_assignment" + Key.make().toString().substring(0, 5) + "_" + parms._model._key.toString() + "_on_" + parms._frame._key.toString();
-        predictions = ((Model.LeafNodeAssignment) parms._model).scoreLeafNodeAssignment(parms._frame, Key.<Frame>make(parms._predictions_name));
+        Model.LeafNodeAssignment.LeafNodeAssignmentType type = null == s.leaf_node_assignment_type ? Model.LeafNodeAssignment.LeafNodeAssignmentType.Path : s.leaf_node_assignment_type;
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ModelMetricsHandler.java]: Should we document this behavior somewhere ?

File: h2o-py/h2o/model/model_base.py [hunk 2]
```diff
@@ -116,17 +116,20 @@ def __repr__(self):
...
 
 
-    def predict_leaf_node_assignment(self, test_data):
+    def predict_leaf_node_assignment(self, test_data, type="Path"):
         """
         Predict on a dataset and return the leaf node assignment (only for tree-based models).
 
         :param H2OFrame test_data: Data on which to make predictions.
+        :param Enum type: How to identify the leaf node. Nodes can be either identified by a path from to the root node
+        of the tree to the node or by H2O's internal node id. One of: ``"Path"``, ``"Node_ID"`` (default: ``"Path"``).
```
Comment (Pscheidl) [inline h2o-py/h2o/model/model_base.py]: Nicely documented default behavior !

File: h2o-core/src/main/java/water/api/ModelMetricsHandler.java (hunk 1 above)
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/ModelMetricsHandler.java]: Customer-facing API could be a little bit more pretty, e.g. `node_id`, `id`, `path`...clients must type `Node_ID` absolutely precisely. `Node_Id` or `node_ID` or even `Node_iD` will result in null value parsed on backend and thus falling back to the default Path. User is not notified about that.

Two options:

- Parse it manually **case-insensitive** in ModelMetricsHandler and let the ModelMetricsSchemaV3 accept just a string. 
- Use a little bit more user-friendly identifiers in the enum.
- Create a separate client-facing Enum with a little bit more user-friendly API.
- Warn the user about non-existing option. The same if the string is empty maybe ?

I'd personally go with options number one or three, if we even consider this to be a problem.

Comment (Pscheidl) [reply h2o-core/src/main/java/water/api/ModelMetricsHandler.java]: To reproduce, just use `  preds3 <- h2o.predict_leaf_node_assignment(model1, train, type = "Node_Id")` in your test and watch the backend falling back to default.

Comment (michalkurka) [reply h2o-core/src/main/java/water/api/ModelMetricsHandler.java]: Fixed

**Pull request merged**